### PR TITLE
Implemented #83 (HTTPHeaders accessors)

### PR DESCRIPTION
**Pull request by amosavian**
This pull request includes static variables for `HTTPVersion` struct. Also implemented easy accessors for `HTTPHeaders` to strict types for header values.
~`Authentication`, `Cookie` and `Set-Cookie` accessors are not implemented yet due to complexity or designing limitations. I will make another PR when they became ready in near future.~

**UPDATE:** All methods are implemented now. Here is mailing list post: mailing list: https://lists.swift.org/pipermail/swift-server-dev/Week-of-Mon-20171113/000713.html

### CODE REVIEW

File: Sources/HTTP/HTTPHeadersAccessors.swift
Comment (helje5) [inline Sources/HTTP/HTTPHeadersAccessors.swift]: This is weird because it can't carry arguments. Maybe this should be called MediaType? And then a ContentType for a content-type?

E.g. how do you do:

    Content-Type: multipart/form-data; boundary=something

or

    Content-Type: text/calendar; component=VCARD

File: Sources/HTTP/HTTPHeadersAccessors.swift
Comment (helje5) [inline Sources/HTTP/HTTPHeadersAccessors.swift]: This needs to be open, an `enum` is not appropriate here for obvious reasons. (at the minimum add a `.other`, but that is not really neat either - rather do constants)

File: Sources/HTTP/HTTPHeadersAccessors.swift
Comment (helje5) [inline Sources/HTTP/HTTPHeadersAccessors.swift]: What about wildcard support? Quality arguments? Supporting this would be the requirement for a structured header:

    Accept: text/html, application/xhtml+xml, application/xml;q=0.9, */*;q=0.8

File: Sources/HTTP/HTTPHeadersAccessors.swift
Comment (helje5) [inline Sources/HTTP/HTTPHeadersAccessors.swift]: If-Match takes an array of etags, supports weak and wildcards. For a typed header this seems very insufficient.

File: Sources/HTTP/HTTPHeadersAccessors.swift
Comment (helje5) [inline Sources/HTTP/HTTPHeadersAccessors.swift]: I'm not convinced `Int64` is right here as you can't expect a x32 implementation to deal w/ x64 Integers. I think that should be rejected at the HTTP level with a proper response.

File: Sources/HTTP/HTTPVersion.swift
Comment (helje5) [inline Sources/HTTP/HTTPVersion.swift:25]: I don't know, maybe, why not. I'm not sure when I would ever use those constants.

File: Sources/HTTP/HTTPHeadersAccessors.swift
Comment (helje5) [inline Sources/HTTP/HTTPHeadersAccessors.swift]: The timezone and locale should at least get cached.

File: Sources/HTTP/HTTPHeadersAccessors.swift
Comment (helje5) [inline Sources/HTTP/HTTPHeadersAccessors.swift]: cache

File: Sources/HTTP/HTTPHeadersAccessors.swift
Comment (helje5) [inline Sources/HTTP/HTTPHeadersAccessors.swift]: the etag header contains an array. also, no support for w/

File: Sources/HTTP/HTTPHeadersAccessors.swift
Comment (helje5) [inline Sources/HTTP/HTTPHeadersAccessors.swift]: escaping?

**Review by helje5 — COMMENTED** [review]
While I'm kinda think that we should have typed headers, I think this PR has tons of issues and doesn't reflect the RFCs.

File: Sources/HTTP/HTTPHeadersAccessors.swift
Comment (amosavian) [reply Sources/HTTP/HTTPHeadersAccessors.swift]: - This property is a getter. setter method(`set(accept:, quality:)`) take quality as a argument, and getter method sorts values. I believe a server implementation doesn't need q value when It's sorted already.
- There is a `ContentType.all` which is wildcard, there are also wildcards for audio, video and text.

Comment (helje5) [reply Sources/HTTP/HTTPHeadersAccessors.swift]: This is intended to be not just for the server, just fwiw.

File: Sources/HTTP/HTTPHeadersAccessors.swift
Comment (amosavian) [reply Sources/HTTP/HTTPHeadersAccessors.swift]: It returns eTags with quotations. so you can have `"W/\"12345\""` as an etag. Maybe we can present value as an associated enum? (e.g. `.strong("12345")`, `.weak("12345")` and `.wildcard("*")`)
We can change design.

File: Sources/HTTP/HTTPHeadersAccessors.swift
Comment (amosavian) [reply Sources/HTTP/HTTPHeadersAccessors.swift]: Apple's `URLResponse.expectedContentLength` is `Int64`, I decided to use Int64 for compatibility. A file can be bigger than 2TB btw.

File: Sources/HTTP/HTTPHeadersAccessors.swift
Comment (amosavian) [reply Sources/HTTP/HTTPHeadersAccessors.swift]: ditto from `If-Match`

File: Sources/HTTP/HTTPHeadersAccessors.swift
Comment (amosavian) [reply Sources/HTTP/HTTPHeadersAccessors.swift]: I implemented `contentCharset` property for `charset` parameter. We can change design btw

File: Sources/HTTP/HTTPHeadersAccessors.swift
Comment (amosavian) [reply Sources/HTTP/HTTPHeadersAccessors.swift]: Encoding should be supported by server. I believe we can't have custom values when they are not supported unless you show me I'm wrong.

File: Sources/HTTP/HTTPHeadersAccessors.swift
Comment (amosavian) [reply Sources/HTTP/HTTPHeadersAccessors.swift]: No, that adds a quotation to newValue if it doesn't have already. for weak etags

File: Sources/HTTP/HTTPHeadersAccessors.swift
Comment (amosavian) [reply Sources/HTTP/HTTPHeadersAccessors.swift]: Yes, I implemented getter and setter method for all headers. They're maybe redundant in some cases.

**Comment by helje5** [comment]
May I also suggest that you take this to the mailing list for discussion. I think we already discussed this kind of stuff and decided not to do typed headers at this level of the framework. It would be part of a real framework on top of this driver.

If we really go the route of typed headers (which I personally kinda like for efficiency), we should actually do that in the storage itself, not just wrapping the String key/value. I wrote some email on that in the list.

**Comment by amosavian** [comment]
@helje5 please check new implementation of EntryTag.

**Comment by amosavian** [comment]
I don't know how to use mailing list unfortunately. I wonder if you could help me.

Headers are not something that you would access frequently. For requests, you may access them once or twice. For response, you set them once. Parsing all request headers when initializing is not an optimized behavior unless they got accessed (lazy). We may introduce some kind of caching later btw.

**Comment by helje5** [comment]
https://lists.swift.org/mailman/listinfo/swift-server-dev

**Comment by amosavian** [comment]
@helje5 I sent email and it's awaiting moderator approval

**Comment by helje5** [comment]
You first need to subscribe to the list with the email address you want to use for sending messages to it. Once this is through, you don't need to wait for moderator approval.

**Comment by amosavian** [comment]
Posted in mailing list: https://lists.swift.org/pipermail/swift-server-dev/Week-of-Mon-20171113/000713.html

File: Sources/HTTP/HTTPHeadersAccessors.swift
Comment (nacho4d) [inline Sources/HTTP/HTTPHeadersAccessors.swift]: I recently read RFC7233 for implementation somewhere else. RCF allows to define any kind of range not only `bytes`. In the specific cases of File Serving `bytes` just happens to be the most common case. Ranges could be used in other REST, etc calls too and it could be `items`, `rows`, etc whatever, restricting this to only bytes might not allow http module users to use this function to other purposes than file serving.

File: Sources/HTTP/HTTPHeadersAccessors.swift
Comment (nacho4d) [inline Sources/HTTP/HTTPHeadersAccessors.swift]: As of RFC7233 and RFC7232 `ifRange` could also be of the form of `Last-Modified` header. instantiating from EntryTag will not allow the later form.

File: Sources/HTTP/HTTPHeadersAccessors.swift
Comment (amosavian) [reply Sources/HTTP/HTTPHeadersAccessors.swift]: Please find fixed RangeType and ifRange implementations

File: Sources/HTTP/HTTPHeadersAccessors.swift
Comment (nacho4d) [inline Sources/HTTP/HTTPHeadersAccessors.swift]: According to RFC7233 range headers can have multiple ranges like: `Range: bytes=3-5,10-15`. So this should be defined as `[Range<Int64>]?`.

Also, the spec for parsing ranges is not as simple as below 4 lines but I guess that can be addressed later as long as the interface is ok :p

Comment (amosavian) [reply Sources/HTTP/HTTPHeadersAccessors.swift]: As far as I understood from examples in RFC, multiple ranges are continuous if we intersect values. e.g:

`bytes=0-0,-1` becomes: -1..<1
`bytes=500-600,601-999` becomes: 500..<1000
`bytes=500-700,601-999` becomes: 500..<1000

Thus it is not an array of range. Please correct me if I'm wrong?

Update: RFC explicitly says ` byte-range-set  = 1#( byte-range-spec / suffix-byte-range-spec )` which means one value is acceptable, and calls multiple ranges as "valid (but not canonical)". All examples are continuous though.

Comment (nacho4d) [reply Sources/HTTP/HTTPHeadersAccessors.swift]: Spec defines various ranges, they do not have to be continuous, for example response in page 11 shows at least two ranges bytes=500-999,7000-7999

As per spec, server CAN merge overlapping or adjacent ranges however that is not a must. Multipart response for partial content is clearly defined which indicates the existence of multiple ranges. 

     HTTP/1.1 206 Partial Content
     Date: Wed, 15 Nov 1995 06:25:24 GMT
     Last-Modified: Wed, 15 Nov 1995 04:58:08 GMT
     Content-Length: 1741
     Content-Type: multipart/byteranges; boundary=THIS_STRING_SEPARATES

     --THIS_STRING_SEPARATES
     Content-Type: application/pdf
     Content-Range: bytes 500-999/8000

     ...the first range...
     --THIS_STRING_SEPARATES
     Content-Type: application/pdf
     Content-Range: bytes 7000-7999/8000

     ...the second range
     --THIS_STRING_SEPARATES--


https://tools.ietf.org/html/rfc7233

Comment (nacho4d) [reply Sources/HTTP/HTTPHeadersAccessors.swift]: Spec defines various ranges, they do not have to be continuous, for example response in page 11 shows at least two ranges bytes=500-999,7000-7999

As per spec, server CAN merge overlapping or adjacent ranges however that is not a must. Multipart response for partial content is clearly defined which indicates the existence of multiple ranges. 

     HTTP/1.1 206 Partial Content
     Date: Wed, 15 Nov 1995 06:25:24 GMT
     Last-Modified: Wed, 15 Nov 1995 04:58:08 GMT
     Content-Length: 1741
     Content-Type: multipart/byteranges; boundary=THIS_STRING_SEPARATES

     --THIS_STRING_SEPARATES
     Content-Type: application/pdf
     Content-Range: bytes 500-999/8000

     ...the first range...
     --THIS_STRING_SEPARATES
     Content-Type: application/pdf
     Content-Range: bytes 7000-7999/8000

     ...the second range
     --THIS_STRING_SEPARATES--


https://tools.ietf.org/html/rfc7233

File: Sources/HTTP/HTTPHeadersAccessors.swift
Comment (nacho4d) [inline Sources/HTTP/HTTPHeadersAccessors.swift]: `bytes` is the most common type and to validate ranges one would get the file size from a file as: `(fileAttributes[FileAttributeKey.size] as? NSNumber)?.uint64Value` because file size are declared as `UInt64` in Foundation. However current declaration is `Int64` which means there would be cases when the string can not be converted to Range<Int64>. How can we address this?. 

Example: try to get a tiny part of a very big file: 
`Range: bytes=18446744073709551600-18446744073709551601`

Are you using `Int64` is because of ranges like: `Range: bytes=-3` which look like negative?. Actually there is no such a thing as negative ranges. In fact they should be converted into a  range object/struct **after** the file size is known, for example if fileSize is `1000` then it would mean: `Range: bytes=997-999` which uses only positive numbers.

I like the idea of having everything statically defined as structs/classes/enums rather than raw strings, but I am not sure having this interface will help to correctly implement the RFC :(

(As a side note, nodejs implementation converts range header string into range objects in a function that receives the fileSize as a parameter.)

Comment (amosavian) [reply Sources/HTTP/HTTPHeadersAccessors.swift]: About Foundation: `HTTPURLResponse.expectedContentLength` and `URLSession` methods return `Int64`.

On negative values, We don't define how to interpret data as we don't know about resource. It's upon higher level framework to convert negative values to meaningful ones or return 416.

About BigInt, Using Decimal will hit performance. We are implementing these method for common cases. A file with that size is not common and higher level framework would have to implement theirs. Should we add limitations to documentation?

Comment (nacho4d) [reply Sources/HTTP/HTTPHeadersAccessors.swift]: OK, if other values are Int64 then I think this should be ok :]
IMO I don't think documentation needs to know about this. Other framework implementers will likely read comments in code

**Comment by amosavian** [comment]
@seabaylea @carlbrown Any chance to get merged? I must begin working on tests.

File: Sources/HTTP/HTTPVersion.swift
Comment (dschwartz783) [reply Sources/HTTP/HTTPVersion.swift]: Maybe someone has an old copy of Netscape lying around, and can't use HTTP 1.1. :(

**Comment by aleksey-mashanov** [comment]
Great work!
Typed headers are definitely the thing a good framework should have. But I have a sense that this too much for an API between an HTTP framework and an HTTP server. Maybe this should be made as a separate module?

I suggest to replace `Int64` with `Int` to follow Swift guidelines:
> Unless you need to work with a specific size of integer, always use `Int` for integer values in your code.

**Comment by amosavian** [comment]
@aleksey-mashanov Please find my response for `Int64` [here](https://github.com/swift-server/http/pull/104#discussion_r152649403).  HTTPURLResponse.expectedContentLength and URLSession methods return `Int64`. NSFileManager returns `UInt64` for file size but `Content-Length` can have negative values. Furthermore it's common to have a file larger than 2GB on a 32-bit system.

Please find my explanation [here](https://gist.github.com/amosavian/084636d1a382717b7d7ce7469aaacc07) on why we should implement typed header in this layer. RFC 5987 is a reason e.g.

**Comment by aleksey-mashanov** [comment]
>  Furthermore it's common to have a file larger than 2GB on a 32-bit system.

Agree. But 32bit server? We still have some in production for historical reasons but it is a pain. In almost every place in a surrounding code this value will be converted to `size_t` (`Int`) and used for generation of dynamic content. I really don't know what is better: precise API or comfortable API. For low-level of this HTTP API maybe first but this PR is out of scope of low-levelness.

> RFC 5987 is a reason e.g.

Agree. For me RFC5987 and parsing of additional parameters (after `;`) are the most interesting parts of this PR. Maybe they should be implemented lower in `HTTPHeaders` as an additional subscripts?
There are a lot of RFCs, a lot of additional headers and even more will arise. It will be hard to follow all of them correctly and most of changes will be source compatibility breaking.

Again. This PR doesn't provide features which are required to efficiently pass data from a framework to a server and back. Nevertheless it is very useful.

**Comment by amosavian** [comment]
> every place in a surrounding code this value will be converted to size_t (Int) 

Swift Foundation's `FileManager` class works with file size as `UInt64`. In case we don't want typecast, it's better to consider `UInt64` rather than `Int`. On the other hand, Swift compiler may give a warning if user tries to cast `UInt64` to `Int64` or `Int`.

Furthermore, `range`, `contentRange` and `contentLength` must have same type imo. `range` must support negative values.

> There are a lot of RFCs, a lot of additional headers and even more will arise.

I mentioned RFC5987 just as an example because it's mentioned in source code as TODO. Definitely There are more cases which I'm not aware of.

>  PR doesn't provide features which are required to efficiently pass data from a framework to a server and back. Nevertheless it is very useful.

This design is implemented as an extension to current design. My primary goal was not to touch current architecture because any change in architecture usually needs extensive workgroup paperworks (PS: I'm not aware how swift-server workgroup works). But my design, which uses initializers for casting instead of functions (except parsing parameters and `q` values), gives the flexibility to change underlying design anytime.

**Comment by amosavian** [comment]
List of optional but unimplemented headers, which are nice to have in portfolio. I won't implement them as a part of this PR:
`Access-Control-*`, `Accept-Patch`, `Alt-Svc`, `Content-Security-Policy`, `DNT`, `Forwarded`, `Proxy-Authenticate`, `Proxy-Authorization`, `Retry-After`, `Tk`, `Warning`, `X-XSS-Protection`

**Comment by amosavian** [comment]
This PR is completely implemented.

**Review by tanner0101 — APPROVED** [review]
awesome

File: Sources/HTTP/HTTPHeadersAccessors.swift
Comment (amosavian) [inline Sources/HTTP/HTTPHeadersAccessors.swift]: Change to `host`

File: Sources/HTTP/HTTPHeadersAccessors.swift
Comment (amosavian) [inline Sources/HTTP/HTTPHeadersAccessors.swift]: Remove this comment

File: Sources/HTTP/HTTPHeadersAccessors.swift
Comment (amosavian) [inline Sources/HTTP/HTTPHeadersAccessors.swift]: Change comment to `host`

File: Sources/HTTP/HTTPHeadersTypes.swift
Comment (amosavian) [inline Sources/HTTP/HTTPHeadersTypes.swift]: utf8 is default encoding, change the order

File: Sources/HTTP/HTTPHeadersTypes.swift
Comment (amosavian) [inline Sources/HTTP/HTTPHeadersTypes.swift]: Fix comment

File: Sources/HTTP/HTTPHeadersTypes.swift
Comment (amosavian) [inline Sources/HTTP/HTTPHeadersTypes.swift]: change type: to `”*”`

File: Sources/HTTP/HTTPHeadersTypes.swift
Comment (amosavian) [inline Sources/HTTP/HTTPHeadersTypes.swift]: remove redundant trimming

File: Sources/HTTP/HTTPHeadersTypes.swift
Comment (amosavian) [inline Sources/HTTP/HTTPHeadersTypes.swift]: ditto

File: Sources/HTTP/HTTPVersion.swift
Comment (amosavian) [reply Sources/HTTP/HTTPVersion.swift:25]: HTTP 1.0 and 1.1 are different only in default values for `Keep-Alive` and `Connection` headers afaik. We will replace instances in current code like [here](https://github.com/swift-server/http/blob/31d066a0e92b4575a4389d592e022fe0b9933956/Sources/HTTP/HTTPStreamingParser.swift#L401)

File: Sources/HTTP/HTTPHeadersAccessors.swift
Comment (amosavian) [inline Sources/HTTP/HTTPHeadersAccessors.swift]: replace with `.method`

File: Sources/HTTP/HTTPHeadersAccessors.swift
Comment (amosavian) [inline Sources/HTTP/HTTPHeadersAccessors.swift]: ditto

File: Sources/HTTP/HTTPHeadersTypes.swift
Comment (amosavian) [inline Sources/HTTP/HTTPHeadersTypes.swift]: Fix comment

File: Sources/HTTP/HTTPHeadersAccessors.swift
Comment (amosavian) [inline Sources/HTTP/HTTPHeadersAccessors.swift]: Handling wildcard

**Comment by mohsen1** [comment]
FYI DOM [`Headers`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers) API allows setting multiple values for the same cookie and ignores the key casing:

```js
var h = new Headers()
h.append('Cookie', 'foo bar')
h.append('cooKie', 'baz')
for (var e of h.entries()) { console.log(e) } 
```

prints:

```
["cookie", "foo bar, baz"]
```

**Comment by ianpartridge** [comment]
Closing as this project is historical, per the README.

**Pull request closed without merging**
